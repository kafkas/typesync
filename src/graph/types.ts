export type MermaidGraphOrientation = 'TB' | 'LR';

export type MermaidGraphLink = [nodeId: string, nodeId: string];

export interface MermaidGraph {
  orientation: MermaidGraphOrientation;
  links: MermaidGraphLink[];
}
