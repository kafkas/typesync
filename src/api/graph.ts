import { GraphGeneration } from '../generators/graph/index.js';
import { objectKeys } from '../util/object-keys.js';
import { GenerateRepresentationResult } from './_common.js';

const SCHEMA_GRAPH_ORIENTATIONS = {
  vertical: true,
  horizontal: true,
};

export type SchemaGraphOrientation = keyof typeof SCHEMA_GRAPH_ORIENTATIONS;

export function getSchemaGraphOrientations() {
  return objectKeys(SCHEMA_GRAPH_ORIENTATIONS);
}

export interface GenerateGraphRepresentationOptions {
  definition: string;
  orientation?: SchemaGraphOrientation;
  debug?: boolean;
}

export interface GenerateGraphOptions extends GenerateGraphRepresentationOptions {
  outFile: string;
  startMarker?: string;
  endMarker?: string;
}

export type GenerateGraphOption = keyof GenerateGraphOptions;

export interface GenerateGraphRepresentationResult extends GenerateRepresentationResult {
  type: 'graph';

  /**
   * A structured representation of the generated graph.
   */
  generation: GraphGeneration;
}

export interface GenerateGraphResult extends GenerateGraphRepresentationResult {}
