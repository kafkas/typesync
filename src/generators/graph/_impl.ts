import { schema } from '../../schema/index.js';
import { assertNever } from '../../util/assert.js';
import type { GraphGeneration, GraphGenerator, GraphGeneratorConfig } from './_types.js';
import { MermaidGraph, MermaidGraphNode, MermaidGraphOrientation } from './mermaid-graph.js';
import { Collection, Document, SchemaGraph } from './schema-graph/interfaces.js';

type SchemaGraphOrientation = 'vertical' | 'horizontal';

// function isGeneric(part: string) {
//   return part.startsWith('{') && part.endsWith('}');
// }

export class GraphGeneratorImpl implements GraphGenerator {
  public constructor(private readonly config: GraphGeneratorConfig) {}

  public generate(s: schema.Schema): GraphGeneration {
    const { documentModels: _ } = s;
    const schemaGraph = this.buildSchemaGraphFromSchema(s);
    const mermaidGraph = this.buildMermaidGraphFromSchemaGraph(schemaGraph);
    return {
      type: 'graph',
      graph: mermaidGraph,
    };
  }

  public buildSchemaGraphFromSchema(_s: schema.Schema): SchemaGraph {
    // TODO: Implement
    return { children: { type: 'literal-graph-children', collections: [] } };
  }

  public buildMermaidGraphFromSchemaGraph(graph: SchemaGraph): MermaidGraph {
    const mermaidGraphOrientation = this.getMermaidGraphOrientation(this.config.orientation);
    const mermaidGraph = new MermaidGraph(mermaidGraphOrientation);
    if (graph.children.type === 'generic-graph-children') {
      this.buildNodeForCollection(mermaidGraph, graph.children.collection);
    } else if (graph.children.type === 'literal-graph-children') {
      graph.children.collections.forEach(col => {
        this.buildNodeForCollection(mermaidGraph, col);
      });
    } else {
      assertNever(graph.children);
    }
    return mermaidGraph;
  }

  private buildNodeForCollection(mermaidGraph: MermaidGraph, col: Collection): MermaidGraphNode {
    const colNode = mermaidGraph.createNode(col.id);
    if (col.children.type === 'generic-collection-children') {
      const docNode = this.buildNodeForDocument(mermaidGraph, col.children.document);
      mermaidGraph.link(colNode, docNode);
    } else if (col.children.type === 'literal-collection-children') {
      col.children.documents.forEach(doc => {
        const docNode = this.buildNodeForDocument(mermaidGraph, doc);
        mermaidGraph.link(colNode, docNode);
      });
    } else {
      assertNever(col.children);
    }
    return colNode;
  }

  private buildNodeForDocument(mermaidGraph: MermaidGraph, doc: Document): MermaidGraphNode {
    const docNode = mermaidGraph.createNode(doc.id);
    if (doc.children !== null) {
      if (doc.children.type === 'generic-document-children') {
        const colNode = this.buildNodeForCollection(mermaidGraph, doc.children.collection);
        mermaidGraph.link(docNode, colNode);
      } else if (doc.children.type === 'literal-document-children') {
        doc.children.collections.forEach(col => {
          const colNode = this.buildNodeForCollection(mermaidGraph, col);
          mermaidGraph.link(docNode, colNode);
        });
      } else {
        assertNever(doc.children);
      }
    }
    return docNode;
  }

  private getMermaidGraphOrientation(orientation: SchemaGraphOrientation): MermaidGraphOrientation {
    switch (orientation) {
      case 'vertical':
        return 'TB';
      case 'horizontal':
        return 'LR';
      default:
        assertNever(orientation);
    }
  }
}

export function createGraphGenerator(config: GraphGeneratorConfig): GraphGenerator {
  return new GraphGeneratorImpl(config);
}
