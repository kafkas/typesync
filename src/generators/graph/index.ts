export { createGraphGenerator } from './_impl.js';
export type * from './_types.js';
export { MermaidGraph, type MermaidGraphNode, type MermaidGraphOrientation } from './mermaid-graph.js';
