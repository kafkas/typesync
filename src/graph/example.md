# Graph

```mermaid
graph TB
    books --> generic_bookId["{bookId}"]
    generic_bookId["Book {bookId}"] --> reviews
    generic_bookId["Book {bookId}"] --> chapters
    generic_bookId["Book {bookId}"] --> translations

    %% authors --> generic_authorId["{authorId}"]
    reviews --> generic_reviewId["Review {reviewId}"]
    chapters --> generic_chapterId["Chapter {chapterId}"]
    translations --> generic_translationId["Translation {translationId}"]

    %% classDef collection fill:#f9f,stroke:#333,stroke-width:2px;
    %% class Books,Authors collection;

    %% classDef document fill:#ccf,stroke:#333,stroke-width:1px;
    %% class Book1,Book2,Author1,Author2 document;

    %% classDef subcollection fill:#cfc,stroke:#333,stroke-width:1px;
    %% class Reviews1,Reviews2 subcollection;

    %% classDef review fill:#cff,stroke:#333,stroke-width:1px;
    %% class Review1_1,Review1_2,Review2_1 review;

```
