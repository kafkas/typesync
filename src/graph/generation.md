```mermaid
graph TD;
    books --> generic_bookId[bookId]
    generic_bookId[bookId] --> reviews

    generic_bookId[bookId] --> chapters

    generic_bookId[bookId] --> translations
```