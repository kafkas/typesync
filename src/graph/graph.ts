import { assertNever } from '../util/assert.js';
import type { Collection, RootCollection } from './nodes.js';
import { MermaidGraph, MermaidGraphLink, MermaidGraphOrientation } from './types.js';

type SchemaGraphOrientation = 'vertical' | 'horizontal';

export function buildMermaidGraph(
  orientation: SchemaGraphOrientation,
  rootCollections: RootCollection[]
): MermaidGraph {
  const links: MermaidGraphLink[] = [];
  rootCollections.forEach(rootCollection => {
    links.push(...buildLinks(rootCollection));
  });
  return { orientation: getMermaidOrientation(orientation), links };
}

function buildLinks(collection: Collection): MermaidGraphLink[] {
  const links: MermaidGraphLink[] = [];
  collection.documents.forEach(document => {
    if (document.type === 'generic-document') {
      const documentNodeId = `generic_${document.genericId}[${document.genericId}]`;
      links.push([collection.id, documentNodeId]);
      document.subCollections.forEach(subCollection => {
        links.push([documentNodeId, subCollection.id]);
      });
    } else if (document.type === 'literal-document') {
      // TODO: Implement
      throw new Error('Unimplemented');
    } else {
      assertNever(document);
    }
  });
  return links;
}

function getMermaidOrientation(orientation: SchemaGraphOrientation): MermaidGraphOrientation {
  switch (orientation) {
    case 'vertical':
      return 'TB';
    case 'horizontal':
      return 'LR';
    default:
      assertNever(orientation);
  }
}
