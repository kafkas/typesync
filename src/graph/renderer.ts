import { StringBuilder } from '@proficient/ds';

import { MermaidGraph } from './types.js';

const TAB = '    ';

export function renderMermaidGraph(graph: MermaidGraph): string {
  const b = new StringBuilder();
  b.append(`graph ${graph.orientation}` + `\n`);
  graph.links.forEach(link => {
    const [leftNodeId, rightNodeId] = link;
    b.append(TAB + leftNodeId + ' --> ' + rightNodeId + `\n`);
  });
  return b.toString();
}
