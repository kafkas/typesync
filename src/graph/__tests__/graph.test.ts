import { GenericDocument, RootCollection, SubCollection, generateMermaidGraph } from '../graph.js';

describe('graph', () => {
  it(`builds graph correctly`, () => {
    const booksCollection = new RootCollection('books', []);
    const genericBookDocument = new GenericDocument('bookId', booksCollection, []);
    booksCollection.documents.push(genericBookDocument);
    const reviewsSubCollection = new SubCollection('reviews', []);
    const chaptersCollection = new SubCollection('chapters', []);
    const translationsCollection = new SubCollection('translations', []);
    genericBookDocument.subCollections.push(reviewsSubCollection, chaptersCollection, translationsCollection);
    const graph = generateMermaidGraph([booksCollection]);

    expect(graph).toMatchSnapshot();
  });
});
