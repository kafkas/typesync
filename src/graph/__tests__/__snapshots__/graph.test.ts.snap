// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph builds graph correctly 1`] = `
"graph TD;
    books --> generic_bookId[bookId]
    generic_bookId[bookId] --> reviews

    generic_bookId[bookId] --> chapters

    generic_bookId[bookId] --> translations

"
`;
